### PR TITLE
fixing interpolation logic

### LINKED ISSUE #499: `interpolated_delta` returns a `OrderError`
Hello 👋 

**Relevant system information:**
 - OS: MacOS 12.4 (running official [pg12-latest](https://hub.docker.com/layers/timescaledb-ha/timescale/timescaledb-ha/pg12-latest/images/sha256-5f08ee3014b5a75fbd971dcbad2ebfa798d6a1f040acc9b0fdcfb4d8712df801?context=explore))
 - PostgreSQL version (output of `SELECT version();`): 12
 - TimescaleDB Toolkit version (output of `\dx timescaledb_toolkit` in `psql`): 1.8.0
 - Installation method: docker

**Describe the bug**

When using `interpolated_delta`, I've got an error:

```
ERROR:  unable to add data to interpolation: OrderError
CONTEXT:  extension/src/counter_agg.rs:129:26
SQL state: XX000
```

Maybe I wrote the query incorrectly? Then query is quite simple though.

**To Reproduce**
1. Run `docker run -p 5432:5432 -e POSTGRES_PASSWORD=password timescale/timescaledb-ha:pg12-latest`
2. Connect to the database
3. Enable the toolkit extension
4. Execute the queries below:

```sql
CREATE TABLE test (
	date timestamp with time zone NOT NULL,
    quantity double precision 
);

INSERT INTO test (date, quantity) VALUES
	('2022-01-01T00:00:00.00+00:00', 10),
	('2022-01-01T00:10:00.00+00:00', 11),
	('2022-01-01T00:20:00.00+00:00', 12),
	('2022-01-01T00:30:00.00+00:00', 16),
	('2022-01-01T00:40:00.00+00:00', 27),
	('2022-01-01T00:50:00.00+00:00', 0),
	('2022-01-01T01:00:00.00+00:00', 3),
	('2022-01-01T01:30:00.00+00:00', 5),
	('2022-01-01T01:50:00.00+00:00', 1),
	('2022-01-01T02:00:00.00+00:00', 6),
	('2022-01-01T02:10:00.00+00:00', 7),
	('2022-01-01T02:30:00.00+00:00', 8),
	('2022-01-01T03:00:00.00+00:00', 5),
	('2022-01-01T03:10:00.00+00:00', 10);
	
SELECT
    date,
    toolkit_experimental.interpolated_delta(
        summary,
        date,
        '30 min',
        LAG(summary) OVER (ORDER by date),
        LEAD(summary) OVER (ORDER by date)
    )
FROM (
    SELECT
        time_bucket('30 min'::interval, date) AS date,
        counter_agg(date, quantity) AS summary
    FROM test
    GROUP BY time_bucket('30 min'::interval, date)
) t;
```

**Expected behavior**
The `SELECT` is expected to return results without error.

**Actual behavior**
An error is returned:
```
ERROR:  unable to add data to interpolation: OrderError
CONTEXT:  extension/src/counter_agg.rs:129:26
SQL state: XX000
```

Thanks!

### PR DESCRIPTION
**Pull request by WireBaron**
This change avoids creating a new start point for an interpolated aggregate if the existing start point is at the interval boundary.

Fixes #499

### CODE REVIEW

File: extension/src/counter_agg.rs [hunk 1]
```diff
@@ -92,7 +92,8 @@ impl<'input> CounterSummary<'input> {
         prev: Option<CounterSummary>,
         next: Option<CounterSummary>,
     ) -> CounterSummary<'static> {
-        let prev = prev.map(|summary| {
+        let prev = if self.first.ts > interval_start {
```
Comment (epgts) [inline extension/src/counter_agg.rs:95]: The test still passes if this is `>=` - can you add a test that would fail in that case?

**Review by epgts — COMMENTED** [review]
I think my comments boil down to "we need more edge-case testing" :)

File: extension/src/time_weighted_average.rs [hunk 2]
```diff
@@ -48,7 +48,7 @@ impl<'input> TimeWeightSummary<'input> {
         prev: Option<TimeWeightSummary>,
         next: Option<TimeWeightSummary>,
     ) -> TimeWeightSummary<'static> {
-        assert!(interval_start <= self.first.ts && interval_start + interval_len >= self.last.ts);
+        assert!(interval_start <= self.first.ts && interval_start + interval_len > self.last.ts);
```
[the remark this answers is not in the document]
Comment (WireBaron) [reply extension/src/time_weighted_average.rs:51]: This is just fixing the assert to reflect that the aggregate is exclusive on the upper bound.  If this assert fails, then the caller is trying to call an interpolated function using an interval that's smaller than their bucketing interval.  It might still be useful to add a test that we handle invalid input correctly, but `pg_test` still has trouble with failure tests.

File: extension/src/time_weighted_average.rs [hunk 3]
```diff
@@ -62,7 +62,7 @@ impl<'input> TimeWeightSummary<'input> {
             _ => self.first
         };
         let new_end = match next {
-            Some(next) if self.last.ts < interval_start + interval_len => {
```
[the remark this answers is not in the document]
Comment (WireBaron) [reply extension/src/time_weighted_average.rs:65]: The removed `if` should never have failed on valid input (and with the assert fixed is guaranteed to not fail).

File: extension/src/gauge_agg.rs [hunk 4]
```diff
@@ -61,11 +61,15 @@ mod toolkit_experimental {
...
-            );
+            let prev = if this.first.ts > interval_start {
+                prev.map(|summary| 
+                    time_weighted_average::TimeWeightMethod::Linear
+                    .interpolate(summary.last, Some(this.first), interval_start)
+                    .expect("unable to interpolate lower bound")
+                ) 
+            } else {
+                None
+            };
```
[the remark this answers is not in the document]
Comment (WireBaron) [reply extension/src/gauge_agg.rs:72]: Yes, copied the counter_agg test here.  We should look at deduplicating this for the next release or the following one (probably as part of stabilizing gauge_agg).

**Review by thatzopoulos — APPROVED** [review]

File: extension/src/counter_agg.rs (hunk 1 above)
Comment (epgts) [reply extension/src/counter_agg.rs:95]: Indeed!  I don't know what confused me, but thanks.

**Review by epgts — APPROVED** [review]
